modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextAreaSkin.java

```diff
@@ -745,52 +724,24 @@ private void paragraphEnd(boolean goPastInitialNewline, boolean select) {
     }
 
     /** {@inheritDoc} */
-    @Override protected PathElement[] getUnderlineShape(int start, int end) {
-        int pStart = 0;
-        for (Node node : paragraphNodes.getChildren()) {
-            Text p = (Text)node;
-            int pEnd = pStart + p.textProperty().getValueSafe().length();
-            if (pEnd >= start) {
-                return p.underlineShape(start - pStart, end - pStart);
-            }
-            pStart = pEnd + 1;
-        }
-        return null;
+    @Override
+    protected PathElement[] getUnderlineShape(int start, int end) {
+        return getTextNode().underlineShape(start, end);
```

Comment:
The former code would return `null` if `start` > `textNode.textProperty().getValueSafe().length()`. I think this is OK, given that the implementation of `Text::underlineShape` checks -- what do you think? This applies to `getRangeShape`, too.

Reply:
I think it's ok.  The javadoc for `TextInputControlSkin::underlineShape()` does not specify a null return value.

In fact, the javadoc for `Text::rangeShape()` and `::underlineShape()` say

`@return an array of {@code PathElement} which can be used to create a {@code Shape}`

implying that it should never be `null` as the return value is typically used in `Shape::getElements().setAll()` or `.addAll()` which will throw an NPE if the thing is null.  In fact, `Text::rangeShape()` (used for underline and range shape) always returns a non-null value.